File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int dfs(int k, vector<vector<int>>& dungeons, vector<bool>& visited, int cnt) {
    int answer = cnt;

    for (int i = 0; i < dungeons.size(); i++) {
        int min_k = dungeons[i][0];
        int use_k = dungeons[i][1];

        if (!visited[i] && k >= min_k) {
            visited[i] = true;

            int result = dfs(k - use_k, dungeons, visited, cnt + 1);

            answer = max(answer, result);

            visited[i] = false;
        }
    }

    return answer;
}

int solution(int k, vector<vector<int>> dungeons) {
    vector<bool> visited(dungeons.size(), false);

    return dfs(k, dungeons, visited, 0);
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def dfs(k, dungeons, visited, cnt):
    answer = cnt
    
    for i in range(len(dungeons)):
        min_k, use_k = dungeons[i]
        
        if not visited[i] and k >= min_k:
            visited[i] = True
            
            result = dfs(k-use_k, dungeons, visited, cnt + 1)
            
            answer = max(answer, result)
            visited[i] = False
            
    return answer
    
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    
    return dfs(k, dungeons, visited, 0)